pkg/controller/volume/attachdetach/attach_detach_controller.go

```diff
@@ -358,18 +380,68 @@ func (adc *attachDetachController) Run(stopCh <-chan struct{}) {
 	go adc.reconciler.Run(stopCh)
 	go adc.desiredStateOfWorldPopulator.Run(stopCh)
 	go wait.Until(adc.pvcWorker, time.Second, stopCh)
+	go wait.Until(adc.vaWorker, time.Second, stopCh)
 	metrics.Register(adc.pvcLister,
 		adc.pvLister,
 		adc.podLister,
 		adc.actualStateOfWorld,
 		adc.desiredStateOfWorld,
-		&adc.volumePluginMgr)
+		&adc.volumePluginMgr,
+		adc.vaLister)
 
 	<-stopCh
 }
 
 func (adc *attachDetachController) populateActualStateOfWorld() error {
 	klog.V(5).Infof("Populating ActualStateOfworld")
+
+	vas, err := adc.vaLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+	volumesToAttach := adc.desiredStateOfWorld.GetVolumesToAttach()
+	if len(volumesToAttach) == 0 {
+		klog.Warning("get empty volumes to attach")
+		//return nil
+	}
+	for _, va := range vas {
+		nodeName := types.NodeName(va.Spec.NodeName)
+		if va.Spec.Source.PersistentVolumeName == nil {
+			continue
+		}
+		var volumeName string
+		driverName := va.Spec.Attacher
+		pluginName := csi.CSIPluginName
+		pvName := *va.Spec.Source.PersistentVolumeName
+		pv, err := adc.pvLister.Get(pvName)
+		var uniqueVolumeName v1.UniqueVolumeName
+		if err != nil && apierrors.IsNotFound(err) ||
+			pv == nil {
+			deleteError := adc.kubeClient.StorageV1().VolumeAttachments().Delete(va.Name, nil)
```

Comment:
Deletion seems like a risky operation. Is it safe in all cases?

Reply:
I have added double check.
if pv does not exist in cache, try to get it from api-server?


wdyt?